### ISSUE TITLE
[HOLD for payment 2023-12-08] [$1000] IOU - Tapping download on manual money request receipt app crashes

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#29788]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.85-0

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

## Action Performed:
1. Launch app
2. Tap fab
3. Tap request money-- Manual
4. Enter an amount
5. Tap next
6. Select a Workspace
7. Tap 3 dots and add the below attached receipt or any image by clicking " choose a document"
8. Tap request
9. Tap on the IOU
10. Tap on the receipt
11. Tap on 3 dots
12. Tap download

## Expected Result:
When user taps download in receipt page, receipt must be downloaded

## Actual Result:
When user taps download in receipt page, app crashes

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

 

https://github.com/Expensify/App/assets/78819774/17ad8c91-7783-4a74-afaf-de469f478c8d

![Bug6240242_1697536533241!w_32904c599d480e967b4c3bba8e072b06c866bb21-2023-10-17_09_38_19 470-2023-10-17_09_41_03 755 (2)](https://github.com/Expensify/App/assets/78819774/8415d643-2bc3-43af-ba8b-1dffeb6697fe)

[logs 3.txt](https://github.com/Expensify/App/files/12931930/logs.3.txt)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01728d6678a4828e28</li>
        <li>Upwork Job ID: 1714368001279598592</li>
        <li>Last Price Increase: 2023-10-30</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Contributor | 27247568</li>
<li>b4s36t4 | Contributor | 27652633</li>
</ul></li>
    </ul>
</details>

**Comment by marcochavezf** [comment]
Posted in [wave4](https://expensify.slack.com/archives/C057F53HR40/p1697560708542619) in case someone has more context about the issue

**Comment by pradeepmdk** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
Tapping download on manual money request receipt app crash

### What is the root cause of that problem?
![Screenshot 2023-10-17 at 9 51 35 PM](https://github.com/Expensify/App/assets/16595846/9326d796-dcc2-4ba6-9847-dc7f67a2087f)

https://github.com/Expensify/App/blob/e92c44f256951dd176e558c9d377cd28ecef1843/src/libs/fileDownload/index.android.js#L46
local file referring to the  `file:/` so `isLocalFile` is false so its pointing the HTTP URL so that its crashing 

### What changes do you think we should make in order to solve the problem?
we can check with  `const isLocalFile = url.startsWith('file:/'); `

**Comment by francoisl** [comment]
@pradeepmdk do you see any recent changes/pull request that could have introduced this crash? All the code you linked to doesn't seem to have changed in a few months, so I'm curious why we're only seeing this now. Or maybe it's because we just started using `AttachmentPicker` in a new component.

Anyway, thanks for the proposal, I'll run some tests to see if it fixes the crash.

**Comment by situchan** [comment]
For deploy blocker issues, offending PR should be pointed out in the root cause of proposal.

**Comment by b4s36t4** [comment]
I just have updated the app from playstore and issue is also there (v1.3.84-10)means this is not a deploy blocker :)

**Comment by situchan** [comment]
@lanitochka17 can you please confirm that this is not reproducible on production?

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
IOU - Tapping download on manual money request receipt app crashes 
### What is the root cause of that problem?

This is an issue with upstream.

Maybe due to slow-network or some other reason the receipt image is pointed to local which causing the issue.

> The RCA for the Crash is correct provided by pradeepmdk i.e accessing `file:/` with only one leading slash.

But the real RCA is from the upstream. As per the documentation when we pass `copyTo` option the library tries to copy the file to temp folder and return the path of the copied file in which the return value from the copy function in upstream is returning `file:` type URI.

https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/io/File.html#toURI()

https://github.com/rnmods/react-native-document-picker/pull/527

https://github.com/rnmods/react-native-document-picker/blob/e1000fd9a93ce3568b1d891917ceae064bed415a/android/src/main/java/com/reactnativedocumentpicker/DocumentPickerModule.java#L315


### What changes do you think we should make in order to solve the problem?

We can upgrade the package `v8.2.1` which uses Uri from android API instead of Java File API 

https://developer.android.com/reference/android/net/Uri#fromFile(java.io.File)


https://github.com/rnmods/react-native-document-picker/blob/5dd2e4063a9806581d99563dce48e1ce2b474264/android/src/main/java/com/reactnativedocumentpicker/DocumentPickerModule.java#L324


### What alternative solutions did you explore? (Optional)
NA

**Comment by b4s36t4** [comment]
I think the issue exists before but we were not able to catch this because we added scan/preview/ menu options feature recently and also I think because of the file-size or some other reason the API request hasn't been completed fully by the time when we open modal so thus it is pointing the image to local path which let the team to catch this issue.

**Comment by marcochavezf** [comment]
I haven't been able to corroborate this issue in production but potentially is not a blocker as @b4s36t4 stated above. Since there are some proposals I will make it external to assign @situchan as C+ (already DM'ed) to move forward.

**Comment by b4s36t4** [comment]
After some-more investigation (after upgrading the package to latest or `8.2.1`) we also need to make another change here in this line https://github.com/Expensify/App/blob/5ad33724f2926f1b687bbac2b3d8fcfced5eb997/src/components/AttachmentModal.js#L184-L195 in which we need to check weather attachment is local or not if local we don't need to add the `encryptedAuthToken` query-param.

After doing these two changes I was able to fix things.

**Comment by situchan** [comment]
I reproduced crash on production

**Comment by marcochavezf** [comment]
Hmm I was able to download the production app and it's working for me on Android with the latest version (1.3.84-10). Probably it's crashing in some android versions

**Comment by b4s36t4** [comment]
Used same version (1.3.84-10) able to re-pro.

**Comment by situchan** [comment]
> Hmm I was able to download the production app and it's working for me on Android with the latest version (1.3.84-10). Probably it's crashing in some android versions

You need to act fast. Please try upload large file (i.e. video) so that local attachment is not replaced until you tap Download button

**Comment by marcochavezf** [comment]
> You need to act fast. Please try upload large file (i.e. video) so that local attachment is not replaced until you tap Download button

Ohh I see, it's an edge case that happens while the file is uploading. In that case, I will remove the deploy blocker label

**Comment by situchan** [comment]
Validating @b4s36t4's proposal. It may take some time to test because android build is very slow

**Comment by pradeepmdk** [comment]
@b4s36t4 @situchan  did you check this test result?
https://github.com/rnmods/react-native-document-picker/pull/527
because in this pull request early this is o/p
```javascript
LOG  [
  {
    "size": 3028,
    "fileCopyUri": "/data/user/0/com.example.reactnativedocumentpicker/cache/7246f3d1-815b-45fa-86ce-a420da059da4/sample.pdf",
    "name": "sample.pdf",
    "type": "application/pdf",
    "uri": "content://com.android.providers.downloads.documents/document/18"
  }
``` 
after merging this pull
```javascript
 LOG  [
  {
    "size": 3028,
    "fileCopyUri": "file:/data/user/0/com.example.reactnativedocumentpicker/cache/d87a83fe-e393-40f2-9f5f-f3ebe2641974/sample.pdf",
    "name": "sample.pdf",
    "type": "application/pdf",
    "uri": "content://com.android.providers.downloads.documents/document/18"
  }
]
``` 
According to wikipedia file:/ and file:/// are the same thing: https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes.3F


and also let me why specific version here `v8.2.1`
because the latest version is `9.0.1`

**Comment by b4s36t4** [comment]
Yes, I just added it as a reference but they have updated their code to support `file://` prefix. It was initial PR which I added to prove my RCA is correct.

**Comment by b4s36t4** [comment]
About version question I also specified latest Or 8.2.1, it can be either. 

Upgrading to a major version may contain breaking changes so suggested any of the version upgrade. 

8.2.1 is the latest release in v8 tree.

**Comment by b4s36t4** [comment]
I have tested my solution and it's working perfectly too. 

> upgraded to 9.0.1

**Comment by pradeepmdk** [comment]
![Screenshot 2023-10-18 at 3 58 18 AM](https://github.com/Expensify/App/assets/16595846/0b254efe-4eaf-480b-b32e-a3ac4a1d3fc7)
@b4s36t4 agree,  latest version correctly return

**Comment by dylanexpensify** [comment]
@situchan any rough eta here?

**Comment by situchan** [comment]
today

**Comment by b4s36t4** [comment]
We might need to hold till PR from @bernhardoj get merged. As that is a required change to make sure the solution works correctly.

**Comment by situchan** [comment]
> We might need to hold till PR from @bernhardoj get merged. As that is a required change to make sure the solution works correctly.

Agree

**Comment by marcochavezf** [comment]
The other [PR](https://github.com/Expensify/App/pull/29738) was merged and deployed to production.

**Comment by isabelastisser** [comment]
Hey @marcochavezf, I'm confused! What are the next steps here? Based on your comment above, it seems like we can move forward with the fix for this issue since the PR for the other issue was deployed. Is that correct? Thanks!

**Comment by b4s36t4** [comment]
@situchan will check and update you.

**Comment by isabelastisser** [comment]
@b4s36t4 @situchan can you please provide an update? Thanks!

**Comment by b4s36t4** [comment]
@isabelastisser been out with sick. Will provide update by trmw

**Comment by situchan** [comment]
@isabelastisser we're still awaiting proposal with successful upload, not only crash

**Comment by dylanexpensify** [comment]
Still waiting on proposals, let's double @isabelastisser?

**Comment by isabelastisser** [comment]
@dylanexpensify Alright, I doubled the price. Waiting for proposals!

**Comment by marcochavezf** [comment]
No overdue, waiting for proposals

**Comment by dylanexpensify** [comment]
Still waiting!

**Comment by dylanexpensify** [comment]
Still pending

**Comment by isabelastisser** [comment]
Waiting for proposals. @dylanexpensify @marcochavezf  do you agree that if we don't have a proposal next week, we should take it internally instead of doubling it to 2K? That's my recommendation.

**Comment by ShogunFire** [comment]
It has been fixed by https://github.com/Expensify/App/pull/30925

**Comment by b4s36t4** [comment]
I don't think that's real fix, we should really update the package.

**Comment by situchan** [comment]
This is still reproducible (android only)
Note: To reproduce, you should select "Choose document", not "Choose from gallery"

I confirmed that "not found page" I am concerned also happens on production. So I don't think this is a new regression from version update of document picker.
This happens on 2nd scan request when re-upload image downloaded from former scan request.

@b4s36t4 let me know if you think the same after test on latest main.
If so, we can go ahead.

**Comment by situchan** [comment]
I finally found the root cause of failing scan request sometimes.
This happens when file name has empty space (i.e. `test receipt.jpg`).
I think this bug is critical and should be fixed for sure but can be out of scope for this GH as this also happens on production app and root cause is different.

We can go with @b4s36t4's [proposal](https://github.com/Expensify/App/issues/29788#issuecomment-1766988153) to fix OP (android crash when download local image).

cc: @marcochavezf 
🎀 👀 🎀 C+ reviewed

**Comment by isabelastisser** [comment]
Hey @marcochavezf, do you approve the proposal above?

**Comment by marcochavezf** [comment]
Thanks @isabelastisser for the heads up, the analysis sounds good, assigning @b4s36t4 🚀

**Comment by dylanexpensify** [comment]
Woot! @b4s36t4 let's get that PR ready!

**Comment by b4s36t4** [comment]
@situchan PR is up https://github.com/Expensify/App/pull/31499.

**Comment by dylanexpensify** [comment]
bump @situchan ^^

**Comment by b4s36t4** [comment]
@dylanexpensify @situchan Asked for a review in the PR, I updated it only yesterday. He might check it today.

**Comment by dylanexpensify** [comment]
Payment upcoming!

**Comment by isabelastisser** [comment]
@situchan please complete the checklist above. Thanks!

**Comment by situchan** [comment]
The bug always existed after we enable local file download. Was issue in upstream.
This was caught by QA team so no need another regression test

**Comment by isabelastisser** [comment]
The payments were made in Upwork. All set!
